# Data preparation
In diesem Notebook wird versucht, aus den Daten der [Strassennamen](https://data.bs.ch/explore/dataset/100189/information/) des Kanton Basel-Stadt die Strassen, welche nach Personen benannt sind, zu extrahieren.

---
## Laden der Daten

In [1]:
url = 'https://data.bs.ch/explore/dataset/100189/download/?format=csv&timezone=Europe/Zurich&lang=de&use_labels_for_header=true&csv_separator=%3B '

Daten in ein pandas-Dataframe abfüllen um eine tabelarische Übersicht über die Daten zu erhalten.
Sollte pandas nicht installiert sein `pip install pandas`

In [2]:
import pandas as pd

df = pd.read_csv(url, sep=';')
df.head()

Id Strasse           Strassenname  \
0         270       Homburgerstrasse   
1         607      Türkheimerstrasse   
2         862  Kleinriehen-Promenade   
3         659         Zürcherstrasse   
4         233            Hardstrasse   

                            Erklärung erste Zeile  \
0      Homburg, Ruine (13. Jh.) bei Läufelfingen,   
1  Turckheim (dt. Türkheim), elsässische Gemeinde   
2                   Kleinriehen, heute Bäumlihof,   
3      Zürich, Hauptort des gleichnamigen Kantons   
4            Hard, mittelalterlicher Flurname für   

                            Erklärung zweite Zeile  \
0  1798 niedergebrannt, ab 1937 wieder restauriert   
1                                              NaN   
2       Landgut und Gartenhaus, erbaut 17.–19. Jh.   
3                                              NaN   
4       stadtnahen Nutzwald des Klosters St. Alban   

                                           Geo Shape  \
0  {"coordinates": [[[7.61278605186816, 47.555152...   
1  {"coordinates": [[[7.574517169860521, 47.55728...   
2  {"coordinates": [[7.617915208772566, 47.569080...   
3  {"coordinates": [[[7.605110252387607, 47.55265...   
4  {"coordinates": [[[7.608488168809716, 47.54850...   

                              Geo Point Erstmals erwähnt Amtlich benannt  \
0  47.555191739157635,7.612705706212208              NaN            1889   
1    47.55899007999238,7.57348436232413              NaN            1877   
2   47.56972353042248,7.620056107528803             1811            1954   
3  47.554425569580275,7.611323562492265             1811            1871   
4   47.55021939912588,7.605043946927342             1695            1860   

  Indextext           Kurztext  
0       HOM      Homburgerstr.  
1       TÜR     Türkheimerstr.  
2       KLE  Kleinriehen-Prom.  
3       ZÜR        Zürcherstr.  
4       HAR           Hardstr.

Nach den Texten suchen, die in `Erklärung erste Zeile` eine Ziffer aufweisen

In [3]:
a = df.query("`Erklärung erste Zeile`.notnull()")
a.head()

Id Strasse           Strassenname  \
0         270       Homburgerstrasse   
1         607      Türkheimerstrasse   
2         862  Kleinriehen-Promenade   
3         659         Zürcherstrasse   
4         233            Hardstrasse   

                            Erklärung erste Zeile  \
0      Homburg, Ruine (13. Jh.) bei Läufelfingen,   
1  Turckheim (dt. Türkheim), elsässische Gemeinde   
2                   Kleinriehen, heute Bäumlihof,   
3      Zürich, Hauptort des gleichnamigen Kantons   
4            Hard, mittelalterlicher Flurname für   

                            Erklärung zweite Zeile  \
0  1798 niedergebrannt, ab 1937 wieder restauriert   
1                                              NaN   
2       Landgut und Gartenhaus, erbaut 17.–19. Jh.   
3                                              NaN   
4       stadtnahen Nutzwald des Klosters St. Alban   

                                           Geo Shape  \
0  {"coordinates": [[[7.61278605186816, 47.555152...   
1  {"coordinates": [[[7.574517169860521, 47.55728...   
2  {"coordinates": [[7.617915208772566, 47.569080...   
3  {"coordinates": [[[7.605110252387607, 47.55265...   
4  {"coordinates": [[[7.608488168809716, 47.54850...   

                              Geo Point Erstmals erwähnt Amtlich benannt  \
0  47.555191739157635,7.612705706212208              NaN            1889   
1    47.55899007999238,7.57348436232413              NaN            1877   
2   47.56972353042248,7.620056107528803             1811            1954   
3  47.554425569580275,7.611323562492265             1811            1871   
4   47.55021939912588,7.605043946927342             1695            1860   

  Indextext           Kurztext  
0       HOM      Homburgerstr.  
1       TÜR     Türkheimerstr.  
2       KLE  Kleinriehen-Prom.  
3       ZÜR        Zürcherstr.  
4       HAR           Hardstr.

In [4]:
dfnumber = a[a["Erklärung erste Zeile"].str.contains('[0-9]')]

In [5]:
dfnumber.head(5)

Id Strasse            Strassenname  \
0          270        Homburgerstrasse   
6          199  General Guisan-Strasse   
10         634         Beim Wasserturm   
14         591            Tanzgässlein   
21         963         Wiesenschanzweg   

                                Erklärung erste Zeile  \
0          Homburg, Ruine (13. Jh.) bei Läufelfingen,   
6               Henri Guisan (1874–1960), General der   
10  Wasserturm, 1925/26 erbaut als Teil der Basler...   
14                            Zum Tanz, 1401 erstmals   
21                  Wiesenschanze, 1799 angelegte und   

                             Erklärung zweite Zeile  \
0   1798 niedergebrannt, ab 1937 wieder restauriert   
6              Schweizer Armee im Zweiten Weltkrieg   
10                                              NaN   
14                               erwähnter Hausname   
21      1856 erneuerte ehemalige Befestigungsanlage   

                                            Geo Shape  \
0   {"coordinates": [[[7.61278605186816, 47.555152...   
6   {"coordinates": [[[7.560156530968611, 47.54856...   
10  {"coordinates": [[[7.592887776795289, 47.52673...   
14  {"coordinates": [[7.587851334546428, 47.559169...   
21  {"coordinates": [[[7.59510707687182, 47.571088...   

                               Geo Point Erstmals erwähnt Amtlich benannt  \
0   47.555191739157635,7.612705706212208              NaN            1889   
6    47.55001731919121,7.565384100053732             1881            1960   
10  47.52675607510471,7.5928652947510695              NaN            1928   
14   47.55926891324485,7.587562251856013             1610            1899   
21  47.572389285986375,7.595065312329401              NaN            1890   

   Indextext             Kurztext  
0        HOM        Homburgerstr.  
6        GUI  General Guisan-Str.  
10       WAS      Beim Wasserturm  
14       TAN         Tanzgässlein  
21       WIE      Wiesenschanzweg

**💡 Erkenntnis**  
Es gibt zu viele Objekt mit einer Jahreszahl

Versuch nach den Texten zu suchen, die in `Erklärung erste Zeile` einen Ziffer und anschliessend ein `-` (En Dash, Unicode: U+2013) aufweisen

In [6]:
dfnumber = a[a["Erklärung erste Zeile"].str.contains('[0-9]–')]
dfnumber.head(5)

Id Strasse            Strassenname  \
6          199  General Guisan-Strasse   
50         502        Rütimeyerstrasse   
51           2  Aeneas Silvius-Strasse   
53         434       Oekolampadstrasse   
58         857        Klingentalgraben   

                                Erklärung erste Zeile  \
6               Henri Guisan (1874–1960), General der   
50        Ludwig Rütimeyer (1825–1895), Professor für   
51               Enea Silvio Piccolomini (1405–1464),   
53  Johannes Oekolampad (1482–1531), Basler Reform...   
58      Klingental, Dominikanerinnenkloster 1274–1557   

                               Erklärung zweite Zeile  \
6                Schweizer Armee im Zweiten Weltkrieg   
50           Anatomie und Zoologie, Evolutionsbiologe   
51  ab 1458 Papst Pius II., Stifter der Universitä...   
53                                                NaN   
58         Graben, Teil der früheren Stadtbefestigung   

                                            Geo Shape  \
6   {"coordinates": [[[7.560156530968611, 47.54856...   
50  {"coordinates": [[[7.575890304203304, 47.54895...   
51  {"coordinates": [[[7.593205357077371, 47.53096...   
53  {"coordinates": [[7.570211275392267, 47.557483...   
58  {"coordinates": [[[7.588554684036573, 47.56315...   

                               Geo Point Erstmals erwähnt Amtlich benannt  \
6    47.55001731919121,7.565384100053732             1881            1960   
50  47.54925418512208,7.5757593685345315              NaN            1897   
51  47.531027360197285,7.593113406541522              NaN            1922   
53   47.55786035014962,7.570172036221308              NaN            1924   
58    47.56363547284315,7.58999466529051             1820            1860   

   Indextext             Kurztext  
6        GUI  General Guisan-Str.  
50       RÜT        Rütimeyerstr.  
51       SIL  Aeneas Silvius-Str.  
53       OEK       Oekolampadstr.  
58       KLI     Klingentalgraben

In [7]:
dfnumber.to_csv("1_StrassenBpotentielMenschzuweisung.csv")

In [8]:
dfnumber.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 147 entries, 6 to 1425
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Id Strasse              147 non-null    int64 
 1   Strassenname            147 non-null    object
 2   Erklärung erste Zeile   147 non-null    object
 3   Erklärung zweite Zeile  119 non-null    object
 4   Geo Shape               147 non-null    object
 5   Geo Point               147 non-null    object
 6   Erstmals erwähnt        33 non-null     object
 7   Amtlich benannt         147 non-null    object
 8   Indextext               147 non-null    object
 9   Kurztext                147 non-null    object
dtypes: int64(1), object(9)
memory usage: 12.6+ KB


----
### Namen extrahieren
Gemäss dp-StrassenBS-Wikidata.ipynb

In [9]:
seperators =["strasse", "-Strasse", "weg", "-Weg", "-Anlage", "anlage", "-Promenade", "rain", "gasse", "gässlein", "gässchen",
             "-Steg", "platz", "-Platz", "-Brücke", "-Passage", "graben", "-Graben", "steg", "-Park", "park", "schanze", "tunnel", "kreisel", "ring", "allee"]

In [10]:
import numpy as np

dfnumberStam = dfnumber.copy() #Sonst gibt es eine Warnung beim setzen der neuen Spalte
dfnumberStam["Stamm"] = dfnumberStam["Strassenname"]
dfnumberStam["suffix"] = np.nan

dfnumberStam.head(3)

Id Strasse            Strassenname  \
6          199  General Guisan-Strasse   
50         502        Rütimeyerstrasse   
51           2  Aeneas Silvius-Strasse   

                          Erklärung erste Zeile  \
6         Henri Guisan (1874–1960), General der   
50  Ludwig Rütimeyer (1825–1895), Professor für   
51         Enea Silvio Piccolomini (1405–1464),   

                               Erklärung zweite Zeile  \
6                Schweizer Armee im Zweiten Weltkrieg   
50           Anatomie und Zoologie, Evolutionsbiologe   
51  ab 1458 Papst Pius II., Stifter der Universitä...   

                                            Geo Shape  \
6   {"coordinates": [[[7.560156530968611, 47.54856...   
50  {"coordinates": [[[7.575890304203304, 47.54895...   
51  {"coordinates": [[[7.593205357077371, 47.53096...   

                               Geo Point Erstmals erwähnt Amtlich benannt  \
6    47.55001731919121,7.565384100053732             1881            1960   
50  47.54925418512208,7.5757593685345315              NaN            1897   
51  47.531027360197285,7.593113406541522              NaN            1922   

   Indextext             Kurztext                   Stamm  suffix  
6        GUI  General Guisan-Str.  General Guisan-Strasse     NaN  
50       RÜT        Rütimeyerstr.        Rütimeyerstrasse     NaN  
51       SIL  Aeneas Silvius-Str.  Aeneas Silvius-Strasse     NaN

In [11]:
for seperator in seperators:

    seriesnumberStam1 = dfnumberStam.Strassenname.str.removesuffix(seperator)
    dfnumberStam2 = pd.DataFrame(seriesnumberStam1)
    dfnumberStam2.rename(columns = {"Strassenname":"temp"}, inplace = True)

    dfnumberStam = pd.concat([dfnumberStam, dfnumberStam2], axis=1)

    dfnumberStam['suffix'] = np.where(dfnumberStam["Strassenname"] != dfnumberStam["temp"], seperator, dfnumberStam["suffix"])
    dfnumberStam['Stamm'] = np.where(dfnumberStam['Strassenname'] != dfnumberStam['temp'], dfnumberStam['temp'], dfnumberStam['Stamm'])

    del dfnumberStam['temp']


In [12]:
dfnumberStam.head()

Id Strasse            Strassenname  \
6          199  General Guisan-Strasse   
50         502        Rütimeyerstrasse   
51           2  Aeneas Silvius-Strasse   
53         434       Oekolampadstrasse   
58         857        Klingentalgraben   

                                Erklärung erste Zeile  \
6               Henri Guisan (1874–1960), General der   
50        Ludwig Rütimeyer (1825–1895), Professor für   
51               Enea Silvio Piccolomini (1405–1464),   
53  Johannes Oekolampad (1482–1531), Basler Reform...   
58      Klingental, Dominikanerinnenkloster 1274–1557   

                               Erklärung zweite Zeile  \
6                Schweizer Armee im Zweiten Weltkrieg   
50           Anatomie und Zoologie, Evolutionsbiologe   
51  ab 1458 Papst Pius II., Stifter der Universitä...   
53                                                NaN   
58         Graben, Teil der früheren Stadtbefestigung   

                                            Geo Shape  \
6   {"coordinates": [[[7.560156530968611, 47.54856...   
50  {"coordinates": [[[7.575890304203304, 47.54895...   
51  {"coordinates": [[[7.593205357077371, 47.53096...   
53  {"coordinates": [[7.570211275392267, 47.557483...   
58  {"coordinates": [[[7.588554684036573, 47.56315...   

                               Geo Point Erstmals erwähnt Amtlich benannt  \
6    47.55001731919121,7.565384100053732             1881            1960   
50  47.54925418512208,7.5757593685345315              NaN            1897   
51  47.531027360197285,7.593113406541522              NaN            1922   
53   47.55786035014962,7.570172036221308              NaN            1924   
58    47.56363547284315,7.58999466529051             1820            1860   

   Indextext             Kurztext           Stamm    suffix  
6        GUI  General Guisan-Str.  General Guisan  -Strasse  
50       RÜT        Rütimeyerstr.       Rütimeyer   strasse  
51       SIL  Aeneas Silvius-Str.  Aeneas Silvius  -Strasse  
53       OEK       Oekolampadstr.      Oekolampad   strasse  
58       KLI     Klingentalgraben      Klingental    graben

#### Mit Wikidata abgleichen

In [ ]:
from SPARQLWrapper import SPARQLWrapper, JSON

In [ ]:
URL und Header für Wikidata-Query erstellen gemäss [Access best practices](https://www.wikidata.org/wiki/Wikidata:Data_access#Access_best_practices)

In [13]:
wdUrl = "https://query.wikidata.org/sparql"
user_agent = 'Streetnamequery/1.0 (https://github.com/CaptainInler/cassda-zertifikatsarbeit)'
sparql = SPARQLWrapper(wdUrl, agent=user_agent)

In [14]:
def queryWd(sparql, subject):
    #print(wdKey)
    query = """
    SELECT ?subject ?subjectLabel ?instanceLabel WHERE {
      ?subject rdfs:label "%s"@de;
               wdt:P31 ?instance.
      SERVICE wikibase:label { bd:serviceParam wikibase:language "de" . }   
    }
    """ % (subject)
    #print(query)
    
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    return sparql

In [15]:
import time, sys
from IPython.display import clear_output

def update_progress(progress):
    bar_length = 20
    if isinstance(progress, int):
        progress = float(progress)
    if not isinstance(progress, float):
        progress = 0
    if progress < 0:
        progress = 0
    if progress >= 1:
        progress = 1

    block = int(round(bar_length * progress))

    clear_output(wait = True)
    text = "Progress: [{0}] {1:.1f}%".format( "#" * block + "-" * (bar_length - block), progress * 100)
    print(text)

In [29]:
from time import sleep

i = 0
total = int(len(dfnumberStam.index))
dfr = pd.DataFrame(columns=["Name", "wikiQLabel", "wikiQ", "instance"])

maxresults = 10

for x in dfnumberStam.index:
    i+=1
    #print(i)
    subject = dfnumberStam['Stamm'][x]
    subjectStr = dfnumberStam['Strassenname'][x]
    #print(f"Subjekt: {subject}")
    sparql = queryWd(sparql, subject)
    update_progress(i / total)
    try:
        results = sparql.query()
        #print(results.info())
        
    except Exception as e:
        print(f"{i} Anfragen bis jetzt")
        #Prüfen auf Statuscode 429 (Too many Requests)
        if e.status == 429:
            print(f'Statuscode 429 aufgetreten: Anfrage geht in {e.headers.get("retry-after")}sec weiter')
            sleep(int(e.headers.get("retry-after"))+2)
        else:
            break
            #raise


    result = results.convert()
    #print(result)

    results_df = pd.json_normalize(result['results']['bindings'])
    #print(results_df)

    if not results_df.empty:
        wikiQ = results_df['subject.value'][0]
        wikiQLabel = results_df['subjectLabel.value'][0]
        instance = results_df['instanceLabel.value'][0]

        #print(f"{subjectStr} | {wikiQLabel}: {wikiQ} -> {instance}")
        dfrtemp = pd.DataFrame([[subjectStr, wikiQLabel, wikiQ, instance]], columns=dfr.columns)
        dfr = pd.concat([dfr, dfrtemp])

    else:
        #print(f"{subjectStr} | {subject}: Kein Eintrag in Wikidata gefunden")
        dfrtemp = pd.DataFrame([[subjectStr, subject, np.nan, np.nan]], columns=dfr.columns)
        dfr = pd.concat([dfr, dfrtemp])

    #print(x)
    #if i > maxresults:
        #break

Progress: [--------------------] 0.7%
1 Anfragen bis jetzt
['_HTTPError__super_init', '__cause__', '__class__', '__context__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', '__suppress_context__', '__traceback__', '__weakref__', '_closer', 'args', 'characters_written', 'close', 'code', 'delete', 'errno', 'file', 'filename', 'filename2', 'fp', 'getcode', 'geturl', 'hdrs', 'headers', 'info', 'msg', 'name', 'reason', 'status', 'strerror', 'url', 'with_traceback']
()
<bound method _TemporaryFileWrapper.close of <HTTPError 403: 'Forbidden'>>


In [ ]:
dfr.head()

In [ ]:
n = dfr.groupby("instance").nunique()

In [ ]:
print(n)

Prüfen wieso in Namen mehr eindeutige Werte stehen als in wikiQLabel und wikiQ

In [ ]:
t = dfr[dfr['instance'] == 'Album']


In [ ]:
t.head()

Von den potenziel 147 Menschen werden "nur" 55 also solche erkannt.  
Ein Problem ist unter anderem, dass Strassen ohne Vor- und Nachnamen überhaupt nicht erkannt werden.  
Versuch auch Personen von Strassen mit nur dem Nachnamen zu treffen.  

In [ ]:
dfnm = dfr[dfr['instance'] != 'Mensch'].reset_index(drop=True)

In [ ]:
def queryWdMenschen(sparql, subject):
    #print(wdKey)
    query = """
    SELECT ?subject ?subjectLabel WHERE {
      ?subject wdt:P734 ?familyname;
               wdt:P31 wd:Q5.
      ?familyname rdfs:label "%s"@de;
      SERVICE wikibase:label { bd:serviceParam wikibase:language "de" . }   
    }
    """ % (subject)
    #print(query)
    
    sparql.setQuery(query)
    sparql.setReturnFormat(JSON)
    return sparql

In [ ]:
from time import sleep

i = 0
total = int(len(dfnm.index))
dfr2 = pd.DataFrame(columns=["2Name", "2wikiQLabel", "2wikiQ"])

maxresults = 10

for x in dfnm.index:
    i+=1
    #print(i)
    subject = dfnm['wikiQLabel'][x]
    subjectStr = dfnm['Name'][x]
    #print(f"Subjekt: {subject}")
    sparql = queryWdMenschen(sparql, subject)
    update_progress(i / total)
    try:
        results = sparql.query()
        #print(results.info())
        
    except Exception as e:
        print(f"{i} Anfragen bis jetzt")
        #Prüfen auf Statuscode 429 (Too many Requests)
        if e.status == 429:
            print(f'Statuscode 429 aufgetreten: Anfrage geht in {e.headers.get("retry-after")}sec weiter')
            sleep(int(e.headers.get("retry-after"))+2)


    result = results.convert()
    #print(result)

    results_df = pd.json_normalize(result['results']['bindings'])
    #print(results_df)

    if not results_df.empty:
        wikiQ = results_df['subject.value'][0]
        wikiQLabel = results_df['subjectLabel.value'][0]

        #print(f"{subjectStr} | {wikiQLabel}: {wikiQ} -> {instance}")
        dfrtemp = pd.DataFrame([[subjectStr, wikiQLabel, wikiQ]], columns=dfr2.columns)
        dfr2 = pd.concat([dfr2, dfrtemp])

    else:
        #print(f"{subjectStr} | {subject}: Kein Eintrag in Wikidata gefunden")
        dfrtemp = pd.DataFrame([[subjectStr, subject, np.nan]], columns=dfr2.columns)
        dfr2 = pd.concat([dfr2, dfrtemp])

    #print(x)
    #if i > maxresults:
        #break

In [ ]:
dfr2.head(10)

In [ ]:
dfnan = dfr2[dfr2['2wikiQ'].isna()]

In [ ]:
dfnan.head(5)

In [ ]:
dfnan.to_csv("2_StrassenBSohneMenschzuweisung.csv")